sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/commands.scala

```diff
@@ -244,34 +244,65 @@ private[sql] case class InsertIntoHadoopFsRelation(
     }
 
     def writeRows(taskContext: TaskContext, iterator: Iterator[InternalRow]): Unit = {
+      // Track which rows have been output to disk so that if a data sort is necessary mid-write,
+      // we don't end up outputting the same data twice
+      val writtenRows: HashSet[InternalRow] = new HashSet[InternalRow]
```

Comment:
Once we make sure that only visit the items once, then the rows will not be outputted twice.